### PR TITLE
Support FunctionVar handlers in EventChain rendering

### LINKED ISSUE #6185: Fix EventChain to separately invoke addEvents for mixed event types
## Background: Event Types in Reflex

There are three kinds of events that can be bound to a component's event trigger (e.g. `on_click`):

1. **EventHandler** — a reference to an `@rx.event` decorated function defined in an `rx.State` subclass. When bound to a trigger, this generates JS code that adds the event (mapping any trigger arguments) to the frontend queue to be sent to the server over the websocket and processed on the backend.

2. **Special frontend events** — built-in events like `_call_script`, `_console_log`, etc. These also generate JS code that adds the event to the frontend event queue, but they are handled in a special frontend path without going to the backend.

3. **EventChainVar** — an escape hatch where a `Var` (representing some expression in the compiled JS) is passed directly as an event trigger. It gets executed inline on the frontend, bypassing all Reflex event loop machinery.

Types 1 and 2 can be "chained" by passing them together in a list to an event trigger prop. Type 3 is special and currently **can only be passed by itself**.

## Problem

Currently, it's not possible to mix normal state event handlers with custom `FunctionVar` event handlers in an event chain:

```python
import reflex as rx

class State(rx.State):
    @rx.event
    def do_a_thing(self):
        print("Doing a thing!")

log_after_timeout = rx.vars.FunctionStringVar.create(
    "(...args) => { setTimeout(() => console.log('Timeout reached!', args), 1000); }"
)

def index():
    return rx.button(
        "Do both",
        on_click=[
            State.do_a_thing,       # EventHandler — needs addEvents()
            log_after_timeout,       # FunctionVar — should be called directly
        ]
    )
```

This fails because `LiteralEventChainVar.create()` tries to pass everything through a single `addEvents([ ... ], args, actions)` call. `FunctionVar` items don't serialize as `ReflexEvent(...)` objects, so they can't go through `addEvents`.

The current workaround is `rx.call_function(log_after_timeout)`, but this is a cludge — it still generates code that routes through `addEvents`.

## Goal

Make EventChain generate a JS function body that handles each item appropriately: `addEvents()` for `EventHandler`/`EventSpec` items, and direct function calls for `FunctionVar` items. The generated code should be a single function that executes each step in sequence.

Currently there are type checks that require a Var to be an `EventChainVar`, but this should be relaxed to allow any `FunctionVar` to be used directly in the chain.

## Example Expected Output

For the example above, the generated JS should look something like:

```javascript
(_event) => {
    addEvents([ReflexEvent("state.do_a_thing", {})], _event, {});
    ((...args) => { setTimeout(() => console.log('Timeout reached!', args), 1000); })(_event);
}
```

Rather than trying to pack both into a single `addEvents([...])` array.

## Acceptance Criteria

- [ ] `LiteralEventChainVar.create()` groups consecutive `EventHandler`/`EventSpec` events into `addEvents()` (`EventChain.invocation`) calls and invokes `FunctionVar`/`EventVar` events as direct function calls
- [ ] The generated JS function body is a sequence of statements (not a single expression), using `explicit_return` or a block body
- [ ] Mixed chains work: `on_click=[State.handler, some_function_var, State.other_handler]` produces 3 separate invocations (addEvents, direct call, addEvents)
- [ ] Pure EventHandler/EventSpec chains (the common case) still produce a single `addEvents()` call — no regression
- [ ] Pure FunctionVar chains produce direct calls without `addEvents`
- [ ] Event actions (preventDefault, stopPropagation) are applied correctly regardless of mix
- [ ] Unit tests covering: pure EventSpec chain, pure FunctionVar chain, mixed chain, mixed chain with event actions

## Key Files

| File | Purpose |
|------|---------|
| `reflex/event.py:2041-2113` | `LiteralEventChainVar.create()` — this is the main code to modify. Currently builds a single `invocation.call(events_array, args, actions)` |
| `reflex/event.py:449-555` | `EventChain.create()` — constructs the EventChain from mixed input types |
| `reflex/vars/function.py:348-403` | `ArgsFunctionOperation` — used to build the JS arrow function. May need `explicit_return=True` or a block-body variant |
| `reflex/vars/function.py:410-459` | `ArgsFunctionOperationBuilder` — builder pattern variant, base of `LiteralEventChainVar` |
| `reflex/constants/compiler.py` | `CompileVars.ADD_EVENTS`, `Hooks.EVENTS`, `Imports.EVENTS` |
| `reflex/utils/format.py` | `format_prop()` — where EventChain is converted to a Var for rendering |

## Notes

- `EventChain.events` is typed as `Sequence[EventSpec | EventVar | EventCallback]` — you'll need to distinguish which items are "addEvents-compatible" (EventSpec/EventHandler/EventVar) vs "direct call" (FunctionVar)
- The `ArgsFunctionOperation` supports `explicit_return=True` which generates `{ return expr }` blocks — you may need a multi-statement block body instead
- Look at how `rx.call_function` currently wraps FunctionVar to understand the existing workaround

### PR DESCRIPTION
**Pull request by FarhanAliRaza**
Allow EventChain to accept frontend FunctionVar handlers alongside EventSpec, EventVar, and EventCallback values.

When a chain contains FunctionVars, keep backend events grouped through addEvents(...) and invoke frontend functions inline with the trigger arguments so mixed chains preserve execution order and DOM event actions like preventDefault and stopPropagation.

Wrap inline arrow functions before emitting VarOperationCall JS so direct invocation renders valid JavaScript, add unit coverage for pure/mixed event-chain formatting and creation, and move upload exception docs to the helper that actually raises them to satisfy darglint.

### All Submissions:

- [x] Have you followed the guidelines stated in [CONTRIBUTING.md](https://github.com/reflex-dev/reflex/blob/main/CONTRIBUTING.md) file?
- [x] Have you checked to ensure there aren't any other open [Pull Requests](https://github.com/reflex-dev/reflex/pulls) for the desired changed?

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

### New Feature Submission:

- [x] Does your submission pass the tests?
- [x] Have you linted your code locally prior to submission?

### Changes To Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

closes #6185

### CODE REVIEW

**Comment by FarhanAliRaza** [comment]
Does it require documentation?

File: reflex/vars/function.py [hunk 1]
```diff
@@ -239,7 +317,13 @@ def _cached_var_name(self) -> str:
         Returns:
             The name of the var.
         """
-        return f"({self._func!s}({', '.join([str(LiteralVar.create(arg)) for arg in self._args])}))"
+        func_expr = str(self._func)
+        if _starts_with_arrow_function(func_expr) and not format.is_wrapped(
```
Comment (masenf) [inline reflex/vars/function.py:321]: do we really need the complex logic to detect an arrow function or can we just unconditionally wrap the expression in parentheses if it's not already wrapped

File: reflex/event.py [hunk 2]
```diff
@@ -2098,17 +2120,80 @@ def create(
...
+                if value.event_actions.get("preventDefault"):
+                    statement_js.append(
+                        "if (_reflex_dom_event?.preventDefault) "
+                        "{_reflex_dom_event.preventDefault();}"
+                    )
+                if value.event_actions.get("stopPropagation"):
+                    statement_js.append(
+                        "if (_reflex_dom_event?.stopPropagation) "
+                        "{_reflex_dom_event.stopPropagation();}"
+                    )
```
Comment (masenf) [inline reflex/event.py]: code for handling the event actions (with the addition of throttle and debounce) already exists in `state.js`, but it's inside the `addEvents` functions. i would suggest pulling this logic out as an exported function in `state.js` that takes the event arg and a function which could either be an `addEvents` partial or some user-passed function.

instead of collecting all the event actions and combining them into the EventChain, it probably is better structure to leave the event actions on the individual EventSpec, and apply any further event actions to the event chain itself. Then use the same function to process these actions at each level.

For example, you might have `applyEventActions(target, event_actions, ...args)`; in this case the `target` could be the entire EventChain function if the chain has its own actions, or the actions could be passed to each `addEvents` call (which internally would call `applyEventActions`).

File: reflex/event.py [hunk 3]
```diff
@@ -482,6 +484,20 @@ def create(
...
+                return value
+            if isinstance(value, FunctionVar):
+                if event_chain_kwargs:
+                    warnings.warn(
+                        f"event_chain_kwargs {event_chain_kwargs!r} are ignored for "
+                        "FunctionVar values.",
+                        stacklevel=2,
+                    )
                 return value
             if isinstance(value, EventVar):
```
Comment (masenf) [inline reflex/event.py]: include FunctionVar here and remove the special case for it above. We should now be treating EventVar and FunctionVar in the same way: creating a new event chain that includes these values in the `events` list so they can be composed and formatted as an EventChainVar later.

File: reflex/event.py [hunk 4]
```diff
@@ -2098,17 +2120,80 @@ def create(
         if invocation is not None and not isinstance(invocation, FunctionVar):
             msg = f"EventChain invocation must be a FunctionVar, got {invocation!s} of type {invocation._var_type!s}."
             raise ValueError(msg)
+        assert invocation is not None
+
+        has_function_var = any(isinstance(e, FunctionVar) for e in value.events)
+
+        if not has_function_var:
```
Comment (masenf) [inline reflex/event.py]: we should strive to reduce special cases in this function. i think it's absolutely fine if we process both mixed and homogenous cases with the same logic.

it's actually better if we have a separate `addEvents` call for each EventSpec because then each of those can have different event actions attached to them and the system is more composable.

File: reflex/event.py [hunk 5]
```diff
@@ -2098,17 +2120,80 @@ def create(
...
+                queueable_group.clear()
+
+            for event in value.events:
+                if isinstance(event, FunctionVar):
+                    flush_queueable_group()
+                    function_call = event.call(*call_args)
+                    statement_js.append(f"{function_call!s};")
+                    statement_var_data.append(function_call._get_all_var_data())
+                else:
+                    queueable_group.append(event)
```
Comment (masenf) [inline reflex/event.py]: just to reiterate, i don't think we need or want to batch up these groups; each non-FunctionVar should get its own `addEvents` invocation.

File: tests/units/test_app.py [hunk 6]
```diff
@@ -1719,23 +1719,40 @@ def test_app_wrap_compile_theme(
             "export function Layout"
         )
     ].strip()
+
+    def _wrap_event_handler_calls(handler_js: str) -> str:
```
Comment (masenf) [inline tests/units/test_app.py]: seems like it might be better to just make these replacements in the asserted string

File: reflex/event.py [hunk 7]
```diff
@@ -483,24 +482,24 @@ def create(
         """
         # If it's an event chain var, return it.
         if isinstance(value, Var):
-            if isinstance(value, EventChainVar):
+            if isinstance(value, LiteralEventChainVar):
```
Comment (masenf) [inline reflex/event.py:492]: is there a reason this didn't stay as `EventChainVar`? seems a bit redundant with the extra case added below

File: reflex/vars/function.py (hunk 1 above)
Comment (FarhanAliRaza) [reply reflex/vars/function.py:321]: It breaks the integration test without this.

Comment (FarhanAliRaza) [reply reflex/vars/function.py:321]: It emitts invalid code like `((typeof)(value))`

**Comment by FarhanAliRaza** [comment]
The failing test is flaky
It can be fixed like this but i dont know if it is right decision.

```
diff --git a/reflex/istate/manager/redis.py b/reflex/istate/manager/redis.py
index 63fc9058..e59d57a8 100644
--- a/reflex/istate/manager/redis.py
+++ b/reflex/istate/manager/redis.py
@@ -153,6 +153,10 @@ class StateManagerRedis(StateManager):
         default_factory=dict,
         init=False,
     )
+    _lock_updates_subscribed: asyncio.Event = dataclasses.field(
+        default_factory=asyncio.Event,
+        init=False,
+    )
     _lock_task: asyncio.Task | None = dataclasses.field(default=None, init=False)
 
     # Whether debug prints are enabled.
@@ -797,8 +801,12 @@ class StateManagerRedis(StateManager):
         }
         async with self.redis.pubsub() as pubsub:
             await pubsub.psubscribe(**handlers)  # pyright: ignore[reportArgumentType]
-            async for _ in pubsub.listen():
-                pass
+            self._lock_updates_subscribed.set()
+            try:
+                async for _ in pubsub.listen():
+                    pass
+            finally:
+                self._lock_updates_subscribed.clear()
 
     def _ensure_lock_task(self) -> None:
         """Ensure the lock updates subscriber task is running."""
@@ -954,6 +962,9 @@ class StateManagerRedis(StateManager):
             lock_id: The ID of the lock.
         """
         token = lock_key.decode().rsplit("_lock", 1)[0]
+        if self._oplock_enabled:
+            self._ensure_lock_task()
+            await self._lock_updates_subscribed.wait()
         if (
             # If there's not a line, try to get the lock immediately.
             not self._n_lock_waiters(lock_key)
@@ -964,8 +975,6 @@ class StateManagerRedis(StateManager):
                     f"{SMR} [{time.monotonic() - start:.3f}] {lock_key.decode()} instaque by {lock_id.decode()}"
                 )
             return
-        # Make sure lock waiter task is running.
-        self._ensure_lock_task()
         async with (
             self._lock_waiter(lock_key) as lock_released_event,
             self._request_lock_release(lock_key, lock_id),

```

File: tests/units/test_app.py (hunk 6 above)
Comment (masenf) [reply tests/units/test_app.py]: seems this function can be removed?

File: reflex/event.py [hunk 8]
```diff
@@ -2098,17 +2118,71 @@ def create(
...
+                _args=FunctionArgs(arg_def),
+                _return_expr=invocation.call(
+                    LiteralVar.create([
+                        LiteralVar.create(event) for event in value.events
+                    ]),
+                    arg_def_expr,
+                    value.event_actions,
+                ),
+                _var_value=value,
+            )
```
Comment (masenf) [inline reflex/event.py]: this special case creates a behavior difference when two EventSpec/EventHandler have different event actions:

```python
import reflex as rx


class State(rx.State):
    @rx.event
    def do_a_thing(self, value: str):
        print(f"Doing a thing! {value}")


log_after_timeout = rx.vars.FunctionStringVar.create("(...args) => { setTimeout(() => console.log('Timeout reached!', args), 1000); }")

def index():
    return rx.vstack(
        rx.button(
            "Do single debounce",
            on_click=[
                State.do_a_thing("first x 1000").debounce(1000),
            ]
        ),
        rx.button(
            "Do two debounce",
            on_click=[
                State.do_a_thing("first x 1000").debounce(1000),
                State.do_a_thing("second x 200").debounce(200),
            ]
        ),
        rx.button(
            "Do two debounce with log",
            on_click=[
                State.do_a_thing("first x 1000").debounce(1000),
                State.do_a_thing("second x 200").debounce(200),
                log_after_timeout.partial("arbitrary args", "can be passed"),
                log_after_timeout,
            ]
        ),
    )

app = rx.App()
app.add_page(index)
```

in the first case, it works as expected, since there is only one debounce action.

in the second case, the debounce actions get rolled up together in the `addEvents` function, which results in a single debounce being called for 200ms. this is a long standing issue we've had in the framework with the composability of event actions.

in the third case, the debounce actions are each applied separately because separate `addEvents` calls are used.

i think if we just get rid of this special case wherein we combine non-FunctionVar events, then we finally fix this bug and allow each event spec to have its own event actions.

i do like how the EventChain actions are applied over the entire function if the chain has its own actions. Currently this isn't really exposed in the framework because the typing does not allow you to directly pass an EventChain to an event trigger prop, but likely in the future we will allow this.

**Comment by masenf** [comment]
> The failing test is flaky It can be fixed like this but i dont know if it is right decision.

made the fix in https://github.com/reflex-dev/reflex/pull/6196

you can merge from `main` into your branch to bring it in

File: reflex/event.py (hunk 7 above)
Comment (FarhanAliRaza) [reply reflex/event.py:492]:
```python

def make_timeout_logger():
    return rx.vars.FunctionStringVar.create(
        "(...args) => { setTimeout(() => console.log('Timeout reached!', args), 1000); }"
    ).to(EventChain)

def test_event_chain_create_wraps_event_chain_typed_function_var_kwargs():
    """FunctionVars cast to EventChain should still compose with chain-level kwargs."""
    frontend_handler = make_timeout_logger()

    chain = EventChain.create(
        frontend_handler,
        args_spec=lambda: (),
        event_actions={"preventDefault": True},
    )

    assert isinstance(chain, EventChain)
    assert chain.events == [frontend_handler]
    assert chain.event_actions == {"preventDefault": True}
```

This case fails without this. `EventChainVar` if is redundant, so I removed that if here https://github.com/reflex-dev/reflex/pull/6188/commits/1d7e57e420c73264aa77a3fd560eec44bd586fcc

Comment (masenf) [reply reflex/event.py:492]: okay sounds good. appreciate the code comment for future readers

File: tests/units/test_event.py [hunk 9]
```diff
@@ -745,23 +760,52 @@ def do_a_thing(self):
...
         args_spec=lambda: (),
-        event_actions={"preventDefault": True},
     )
     rendered = str(LiteralVar.create(chain))
 
     assert "applyEventActions(" not in rendered
-    assert rendered.count("addEvents(") == 1
+    assert rendered.count("addEvents(") == 2
+    assert '["debounce"] : 1000' in rendered
+    assert '["debounce"] : 200' in rendered
```
Comment (masenf) [inline tests/units/test_event.py]: probably also assert that these each appear once in the rendered output

File: tests/units/test_event.py [hunk 10]
```diff
@@ -745,23 +760,52 @@ def do_a_thing(self):
...
+        args_spec=lambda: (),
+        event_actions={"preventDefault": True},
+    )
+    rendered = str(LiteralVar.create(chain))
+
+    assert "applyEventActions(" in rendered
+    assert rendered.count("addEvents(") == 2
+    assert '["debounce"] : 1000' in rendered
+    assert '["debounce"] : 200' in rendered
     assert '["preventDefault"] : true' in rendered
```
Comment (masenf) [inline tests/units/test_event.py]: similarly here, lets assert that each of these action strings appear exactly once in the code.

**Comment by masenf** [comment]
@FarhanAliRaza I added some commits to catch some additional cases that were missing while i was testing manually.

Specifically code like this:

```python
        rx.form(
            rx.input(
                name="input_value",
                on_blur=lambda v: rx.EventChain.create(
                    value=log_after_timeout.partial(v.to(str).capitalize()),
                    args_spec=lambda: (),
                ),
            ),
            rx.button(
                "Submit",
            ),
            on_submit=[
                log_after_timeout,
                lambda fd: [
                    log_after_timeout.partial(fd.to(dict)["input_value"]),
                    rx.console_log(fd.to(dict)["input_value"]),
                ],
                State.on_submit,
            ],
        ),
```

2 behavior changes:
- When a FunctionVar is passed, the arguments dictated by the args_spec for the event trigger must be applied as a `.partial` so the handler gets the transformed args 
- When a lambda is passed in the chain, it should be able to return the same types of heterogeneous values that are accepted by the chain itself.

Additionally, lambda return value validation was consolidated into `call_event_fn`.

Can you explore more edge cases around this and add some additional test cases to the unit tests suite.

**Comment by masenf** [comment]
test failure was from another interesting case that we don't have a unit test for yet

```python
            on_key_down=lambda key: rx.cond(
                rx.Var.create(["a", "s", "d", "w"]).contains(key),
                GlobalHotkeyState.update_key(key),
                rx.console_log(key)
            )
```

here the event handler is a lambda that returns some conditional code that runs on the frontend before deciding whether to trigger a backend event or do something else.

Stuff like this still doesn't work though

```python
            rx.input(
                name="input_value",
                on_blur=lambda v: rx.cond(
                    v == "foo",
                    log_after_timeout.partial("Input was foo!"),
                    State.do_a_thing(v),
                ),
            ),
```

When we have an `rx.cond` that returns heterogenous values (either FunctionVar or EventVar), then that gets rejected by the new lambda return type checking logic. Idk if we want to allow that to return any arbitrary var, or if requiring the user to say `rx.cond(...).to(rx.event.EventVar)` is good enough for now. I'm probably more inclined to the latter.

**Comment by FarhanAliRaza** [comment]
I fixed it locally by adding this.

Is it correct?


```python

 elif isinstance(e, Var) and safe_issubclass(
            e._var_type, (EventSpec, EventChain)
        ):
            # Helpers like rx.cond() can return event-typed Vars that should be
            # treated the same as literal EventSpec/EventChain values here.
            e = e.guess_type()
            
```


Update:
But now that integeration test passes so not pushing this.

**Comment by masenf** [comment]
> When we have an rx.cond that returns heterogenous values (either FunctionVar or EventVar), then that gets rejected by the new lambda return type checking logic. Idk if we want to allow that to return any arbitrary var, or if requiring the user to say rx.cond(...).to(rx.event.EventVar) is good enough for now. I'm probably more inclined to the latter.

I think we have some special logic somewhere that is treating the `rx.cond` case specially. Even with the explicit cast to EventVar, i'm not able to get it to work because it wants to treat the whole thing as an EventVar (and thus wrap it in `addEvents`), which just doesn't work if one of the branches of the `rx.cond` is not a real EventSpec.

I think we can just defer on this problem for now; it wasn't working before and still doesn't work now. We could improve it in the future.

File: reflex/event.py [hunk 11]
```diff
@@ -2098,17 +2108,55 @@ def create(
...
+                    {},
+                )
+            )
+            for event in value.events
+        ]
+
+        if not statements:
+            statements.append(invocation.call(LiteralVar.create([]), arg_def_expr, {}))
+
+        statement_var_data = [statement._get_all_var_data() for statement in statements]
```
Comment (masenf) [inline reflex/event.py]: i think this might be the line that is causing the CodSpeed regression.

usually if you just combine the vars, the parser will pick up any nested VarData automatically:

```python
            statement_block = Var(
                _js_expr=f"{{{''.join(f'{statement};' for statement in statements)}}}",
            )
```

i would expect the above to be equivalent to how it was written without the explicit VarData merge

Comment (FarhanAliRaza) [reply reflex/event.py]: You were right, it is fixed now.

**Comment by masenf** [comment]
This PR looks good and works well. I would just like to see a couple more test cases:

* Pass FunctionVar with a non-empty args spec should apply the spec arguments partially
  * Lambda function that returns FunctionVar should NOT get arg spec applied
* Lambda function should be able to return sequence of mixed type events
* Lambda function should be able to return EventChain explicitly -- this should go through unmodified
